python/ray/serve/api.py

```diff
@@ -571,3 +577,102 @@ def delete(name: str, _blocking: bool = True):
+    the multiplexed function will unload the model with LRU policy.
+
+    If you want to release resource after the model is loaded, you can define
+    a `__del__` method in your model class. The `__del__` method will be called when
+    the model is unloaded.
+
+    Example:
+
+    .. code-block:: python
+            from ray import serve
+
+            @serve.deployment
+            class MultiplexedDeployment:
+
+                @serve.multiplexed(num_models_per_replica=5)
+                def load_model(self, model_id: str) -> Any:
+                    # Load model with the given tag
+                    # You can use any model loading library here
+                    # and return the loaded model. load_from_s3 is
+                    # a placeholder function.
+                    return load_from_s3(model_id)
+
+                async def __call__(self, request):
+                    # Get model id from request context
+                    model_id = serve.get_model_id()
+                    # Load model with the given model id
+                    model = await self.get_model(tag)
+                    # Call the model
+                    return model(request)
+
+
+    Args:
+        num_models_per_replica: number of models to be loaded on each replica.
+        By default, it is 0, which means all models will be loaded on all replicas.
+    """
+    if func is not None:
+        if not callable(func):
+            raise TypeError(
+                "@serve.multiplex can only be used to decorate functions or methods."
+            )
+
+    def _multiplex_decorator(func):
+
+        signature = inspect.signature(func)
+        if len(signature.parameters) == 0:
+            raise ValueError(
+                "@serve.multiplex can only be used to decorate functions or methods "
+                "with at least one 'model_id: str' argument."
+            )
+
+        @wraps(func)
+        async def _multiplex_wrapper(*args, **kwargs):
+            self = _extract_self_if_method_call(args, func)
+
+            if self is None:
+                multiplex_object = func
+            else:
+                multiplex_object = self
+                args = args[1:]
+            multiplex_attr = f"__serve_multiplex_{func.__name__}"
+            if not hasattr(multiplex_object, multiplex_attr):
+                model_multiplex_wrapper = _ModelMultiplexWrapper(
+                    func, self, num_models_per_replica
+                )
+                setattr(multiplex_object, multiplex_attr, model_multiplex_wrapper)
+            else:
+                model_multiplex_wrapper = getattr(multiplex_object, multiplex_attr)
+            return await model_multiplex_wrapper.load_model(*args, **kwargs)
+
+        return _multiplex_wrapper
+
+    return _multiplex_decorator(func) if callable(func) else _multiplex_decorator
+
+
+@PublicAPI(stability="alpha")
+def get_model_id() -> str:
```

Comment:
Do we need to support this as a public API?

Reply:
@sihanwang41 can we call this `get_multiplexed_model_id()` to make it clear it's used for multiplexing?